module/ice-9/generic-hash-tables.scm

```diff
@@ -0,0 +1,854 @@
+is set to the result of calling FAILURE and the new value is returned."
+  (assert-mutable ht)
+  (with-hashx-values (h a real-table) ht
+    (if (ht-weakness ht)
+        ;; Separate the case where ht is weak - don't use handle
+        (let* ((value (hashx-ref h a real-table key ht-unspecified)))
+          (cond ((eq? ht-unspecified value)
+                 (let ((value (failure)))
+                   (hashx-set! h a real-table key value)
+                   value))
+                (else value)))
+        (let ((handle (hashx-create-handle! h a real-table key ht-unspecified)))
+          (if (eq? ht-unspecified (cdr handle))
+              (set-cdr! handle (failure)))
+          (cdr handle)))))
+
+(define (hash-table-intern!/default ht key default)
+  "Effectively invokes HASH-TABLE-REF with the given arguments and
+returns what it returns. If KEY was not found in hash-table, its value
+is set to DEFAULT and DEFAULT is returned."
+  (assert-mutable ht)
+  (with-hashx-values (h a real-table) ht
+    (if (ht-weakness ht)
+        ;; Separate the case where ht is weak - don't use handle
+        (let* ((value (hashx-ref h a real-table key ht-unspecified)))
+          (cond ((eq? ht-unspecified value)
+                 (hashx-set! h a real-table key default)
+                 default)
+                (else value)))
+        (let ((handle (hashx-create-handle! h a real-table key default)))
+          (cdr handle)))))
+
+(define* (hash-table-update! ht key updater #:optional
+                             (failure (lambda () (error "Key not in table" key ht)))
+                             success)
+  "Semantically equivalent to, but may be more efficient than, the
+ following code:
+ (HASH-TABLE-SET! HT KEY (UPDATER (HASH-TABLE-REF HT KEY [FAILURE [SUCCESS]])))
+
+Signals an error if HT is immutable. Otherwise, if KEY is found in hash
+table HT, its associated VALUE is set to (UPDATER (SUCCESS VALUE)),
+or (UPDATER VALUE) if SUCCESS isn't provided. If KEY is not found, sets
+the new value to the result of (UPDATER (FAILURE)) if FAILURE is
+provided, or signals an error otherwise."
+  (assert-mutable ht)
+  (with-hashx-values (h a real-table) ht
+    (if (ht-weakness ht)
+        ;; Separate the case where ht is weak - don't use handle
+        (let* ((old (hashx-ref h a real-table key ht-unspecified))
+               (new (updater (if (eq? old ht-unspecified)
+                                 (failure)
+                                 (if success (success old) old)))))
+          (hashx-set! h a real-table key new))
+        (let ((handle (hashx-get-handle h a real-table key)))
+          (cond (handle
+                 (let* ((old (cdr handle))
+                        (new (updater (if success (success old) old))))
+                   (set-cdr! handle new)))
+                (else
+                 (let ((new (updater (failure))))
+                   (hashx-set! h a real-table key new)))))))
+  *unspecified*)
```

Comment:
I think I remember Mark saying `*unspecified*` should not be used.

Reply:
It's a global exported value, if it shouldn't be used, it should be internal. Plus, old SRFI-69 code already used that and 'tests/r6rs-hashtables.test' fails if these procedures don't return *unspecified*.